### ISSUE TITLE
Update asyncbigtable to 0.3.2 with 1.0.0 Bigtable client library

### ISSUE THREAD

**Issue by goll** [OpenTSDB/asyncbigtable#32]
As per @sduskis latest commits we now have the final 1.0.0 Bigtable client library dependency in pom.xml:

```xml
<dependency>
  <groupId>com.google.cloud.bigtable</groupId>
  <artifactId>bigtable-hbase-1.x</artifactId>
  <version>1.0.0</version>
</dependency>
```
Can we now properly close https://github.com/OpenTSDB/asyncbigtable/issues/24 by releasing an updated 0.3.2 asyncbigtable?

Currently the asyncbigtable version in OpenTSDB 2.4.0RC2 is broken (only 0.3.0 is uploaded):
https://github.com/OpenTSDB/opentsdb/blob/v2.4.0RC2/third_party/asyncbigtable/include.mk#L16

cc: @tsuna @mbrukman @manolama

**Comment by sduskis** [comment]
I'm happy to release a 0.3.2 version with this change and the other bug fixes.  We have more radical change we want to make a bit later, so it does seem like we're currently at a good point to make a release.

**Comment by sduskis** [comment]
Cloud Bigtable 1.2.0 client released with an integration of HBase 2.x's new Async API.  We have a prototype that does real async calls.  Hopefully, we'll have PRs for it by some time next week.

**Comment by sduskis** [comment]
@goll: What do we need to make a new release?  We did quite a bit of work to update the client.

**Comment by mbrukman** [comment]
@csoulios, can you please make a new release of the `asyncbigtable` module and update the main OpenTSDB repo to use the latest version? Thanks!

**Comment by goll** [comment]
@sduskis @mbrukman Sorry for the delay, currently in the process of building/testing the latest asyncbigtable with OpenTSDB v2.4.0RC2, just have one question about `<artifactId>bigtable-hbase-2.x</artifactId>` considering this OpenTSDB [directive](http://opentsdb.net/docs/build/html/user_guide/backends/bigtable.html?highlight=hbase.client.connection.impl): should the `pom.xml` use `bigtable-hbase-1.x` together with `com.google.cloud.bigtable.hbase1_2.BigtableConnection`?

Or is `bigtable-hbase-2.x` backwards compatible with `com.google.cloud.bigtable.hbase1_2.BigtableConnection`?

Otherwise we should revert the artifactId to 1.x and just use that.

**Comment by csoulios** [comment]
Hi @mbrukman I will do the release

**Comment by sduskis** [comment]
@goll: Thanks for following up.

*TL;DR; I would prefer to update the docs and use HBase 2.x.*

### HBase 1.x is fundamentally broken for async
 
At a high level, the HBase 1.x API is synchronous.  HBase 2.x has a synchronous API.  The use of the synchronous HBase 1.x API will have the following drawbacks:

1) **AsyncBigtable using HBase 1.x is synchronous for the most part.**  It's quite confusing to users.

2) **AsyncBigtable using HBase 1.x is error prone and fails silently.**  Using a synchronous API under the covers doesn't allow users of AsynchBigtable to know when there's a failure.

### `com.google.cloud.bigtable.hbase1_2.BigtableConnection`

Users no longer have to set the `Connection`.  I used `BigtableConfiguration.configure()` that automatically pulls in the correct `Connection` that's specified in the version of the jar that's used.

### Updating docs
Once the release is made, I will need to follow up on how to modify the docs, or modify the docs before the release, if required.

**Comment by goll** [comment]
@sduskis: Thanks for the explanation, exactly what I was wondering!

So the `hbase.client.connection.impl` directive can be removed from the OpenTSDB configuration entirely in that case since asyncbigtable built with hbase-2.x will automatically pull the correct config from the jar?

**Comment by sduskis** [comment]
That's correct; `hbase.client.connection.impl` can be removed from the configuration.

**Comment by goll** [comment]
@sduskis using a manually built asyncbigtable from latest master and removing the OpenTSDB v2.4.0RC2 directive resulted in the following exception on start:
```
ERROR [grpc-default-executor-1] HBaseClient: BigTable API: CompletableFuture failed with exception
java.util.concurrent.CompletionException: java.lang.NullPointerException
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273) ~[na:1.8.0_171]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280) ~[na:1.8.0_171]
	at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:824) ~[na:1.8.0_171]
	at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:797) ~[na:1.8.0_171]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474) ~[na:1.8.0_171]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977) ~[na:1.8.0_171]
	at com.google.cloud.bigtable.hbase2_x.FutureUtils$2.onFailure(FutureUtils.java:51) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at com.google.common.util.concurrent.Futures$6.run(Futures.java:1310) ~[guava-18.0.jar:na]
	at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:457) ~[guava-18.0.jar:na]
	at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156) ~[guava-18.0.jar:na]
	at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145) ~[guava-18.0.jar:na]
	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:202) ~[guava-18.0.jar:na]
	at com.google.cloud.bigtable.grpc.async.AbstractRetryingOperation$GrpcFuture.setException(AbstractRetryingOperation.java:119) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at com.google.cloud.bigtable.grpc.async.AbstractRetryingOperation.setException(AbstractRetryingOperation.java:271) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at com.google.cloud.bigtable.grpc.async.AbstractRetryingOperation.onError(AbstractRetryingOperation.java:222) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at com.google.cloud.bigtable.grpc.async.AbstractRetryingOperation.onClose(AbstractRetryingOperation.java:187) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at com.google.cloud.bigtable.grpc.io.ChannelPool$InstrumentedChannel$2.onClose(ChannelPool.java:210) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:41) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at com.google.cloud.bigtable.grpc.io.RefreshingOAuth2CredentialsInterceptor$UnAuthResponseListener.onClose(RefreshingOAuth2CredentialsInterceptor.java:85) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:41) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:684) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:41) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:391) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:475) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:63) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:557) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$600(ClientCallImpl.java:478) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:590) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_171]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_171]
	at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_171]
Caused by: java.lang.NullPointerException: null
	at com.google.cloud.bigtable.hbase.adapters.admin.TableAdapter.adapt(TableAdapter.java:96) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at com.google.cloud.bigtable.hbase2_x.adapters.admin.TableAdapter2x.adapt(TableAdapter2x.java:92) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at com.google.cloud.bigtable.hbase2_x.BigtableAsyncAdmin.lambda$getDescriptor$20(BigtableAsyncAdmin.java:273) ~[asyncbigtable-0.3.1-SNAPSHOT-jar-with-dependencies.jar:na]
	at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:822) ~[na:1.8.0_171]
	... 30 common frames omitted
```
Let me know if I can help by testing out further.

**Comment by sduskis** [comment]
Thanks.  I'll take a look.

**Comment by sduskis** [comment]
@goll, did you set a project and instance?

**Comment by goll** [comment]
@sduskis the only change in opentsdb.conf was the removal of the `hbase.client.connection.impl` directive, these directives are still correctly defined:
```
google.bigtable.auth.json.keyfile
google.bigtable.auth.service.account.enable
google.bigtable.grpc.channel.count
google.bigtable.project.id
google.bigtable.instance.id
```

**Comment by sduskis** [comment]
@goll: Here's [TableAdapter.java:96](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/bigtable-client-1.3.0/bigtable-hbase-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/admin/TableAdapter.java#L96).  I'm trying to figure out what could have been null.  It looks like the Cloud Bigtable client needs a bit more checking for null pointers.

The problem likely relates to a bug in the underlying client when tables don't exist.  It's supposed to throw a `TableNotFoundException`, but right now it looks like it throws a NPE.  I'll do what I can to fix the underlying HBase client, and then I'll test that case with AsyncBigtable.

**Comment by sduskis** [comment]
@goll: what do I need to do to replicate the problem you saw?  What functionality did you test?

**Comment by goll** [comment]
@sduskis you can use this to build an image that should reproduce the problem: https://github.com/goll/opentsdb-bigtable/tree/latest

The only tested operation was an attempt to apply a change to the k8s deployment using the new image with the latest asyncbigtable jar which failed on starting the containers with the previously posted error log^.

**Comment by sduskis** [comment]
@goll: We tracked down a bug on the Cloud Bigtable client and the way it handled error cases for the methods used by `HBaseClient.ensureTableFamilyExists()`.  I submitted #46 to upgrade the version of the Cloud Bigtable client used by asyncbigtable which fixes our error handling.

This didn't actually fix the problem you saw, but we'll be able to get better diagnostics on the actual problem that you're experiencing.

**Comment by sduskis** [comment]
@goll, just to be clear, can you please attempt another release?

**Comment by sduskis** [comment]
@goll or @csoulios, can we please try to make another release?

**Comment by csoulios** [comment]
@sduskis you want me to release 0.3.1 from existing master branch?

**Comment by sduskis** [comment]
Can you please release a 0.4.0 from master?  The changes we made deserve a minor semVer change.

**Comment by csoulios** [comment]
@sduskis @goll just released v0.4.0. 
https://oss.sonatype.org/content/repositories/releases/com/pythian/opentsdb/asyncbigtable/0.4.0/
Let me know if it's ok.
